src/anyio/_backends/_asyncio.py

```diff
@@ -499,28 +499,25 @@ async def __aexit__(
         while self.cancel_scope._tasks:
             try:
                 await asyncio.wait(self.cancel_scope._tasks)
-            except asyncio.CancelledError:
+            except asyncio.CancelledError as exc:
+                # This task was cancelled natively; reraise the CancelledError later
+                # unless this task was already interrupted by another exception
                 self.cancel_scope.cancel()
+                if exc_val is None:
+                    exc_val = exc
+                    ignore_exception = False
+                    self._exceptions.append(exc)
 
         self._active = False
         if self._exceptions:
-            exc: BaseException | None
-            group = BaseExceptionGroup("multiple tasks failed", self._exceptions)
-            if not self.cancel_scope._parent_cancelled():
-                # If any exceptions other than AnyIO cancellation exceptions have been
-                # received, raise those
-                _, exc = group.split(is_anyio_cancelled_exc)
-            elif all(is_anyio_cancelled_exc(e) for e in walk_exception_group(group)):
-                # All tasks were cancelled by AnyIO
-                exc = CancelledError()
-            else:
-                exc = group
-
-            if isinstance(exc, BaseExceptionGroup):
-                exc = collapse_exception_group(exc)
-
-            if exc is not None and exc is not exc_val:
-                raise exc
+            group = BaseExceptionGroup(
+                "unhandled errors in a TaskGroup", self._exceptions
+            )
+            _, errors = group.split(CancelledError)
+            if errors is not None:
+                raise errors
+            elif exc_val is not None and not ignore_exception:
+                raise exc_val
```

Comment:
This is the follow-up for "reraise the CancelledError later unless this task was already interrupted by another exception" right? As well as if `exc_val` is not null on context exit? Perhaps a comment about these cases (particularly when the second case would occur) would be helpful.

Reply:
This was just about the hardest part to figure out, as AnyIO is treading into uncharted territory as far as cancellation semantics go. I need to double check that I got this right, or at least that I've made it behave consistently.

Reply:
IIRC the idea was to raise a single cancellation exception if the host task was cancelled and there were no other errors.

Reply:
At least `asyncio.TaskGroup` behaves this way.

Reply:
In the case where `exc_val` is the exception originally pased to `__aexit__`, you will get a nicer traceback (no extraneous frame) if you `return False` rather than explicitly reraising it.

Reply:
In that case the test here fails and the execution proceeds to the end of the method where we do exactly that.

Reply:
I disagree. Imagine that there are no child tasks, and the nested child is cancelled by `Task.cancel()` rather than `CancelScope.cancel()`. Then you wind up with `_exceptions` containing one CancelledError, no `errors`, `exc_val is not None`, and `not ignore_exception` (because the cancellation wasn't for this task group's cancel scope). You will explicitly reraise the CancelledError when you should be implicitly reraising it by returning False.

Reply:
I don't think I understand the scenario you're describing here. Could you illustrate with some sample code?

Reply:
```
async def main():
  async with anyio.create_task_group() as tg:
    asyncio.get_running_loop().call_later(1, asyncio.current_task().cancel)
    await anyio.sleep(5)
asyncio.run(main())
```
I contend that the resulting traceback will have a stack frame from `TaskGroup.__aexit__`. It doesn't need to, and on Trio it wouldn't. It's a minor polish thing, not a big deal, but under the context manager protocol, if you want to have the same exception raised out of the context manager as reached it, then you should return False rather than raising anything.

Reply:
I think I remembered a different implementation of `__aexit__()` in which I didn't add `exc_val` into `self._exceptions`. But IIRC that had different issues. Now the issue is that the code in the `if self._exceptions:` block has no way to determine whether it should reraise `exc_val` (if it was caught inside `__aexit__()`) or to return `False` (if the context manager was exited with it).

Reply:
I didn't mean this wasn't fixable. What I left unsaid was that I would need to add an extra local variable just for distinguishing between these two cases.

Reply:
It might be clearer not to reassign `exc_val` anyway. That'd make the case I had questions about at https://github.com/agronholm/anyio/pull/586#discussion_r1263933896 more obvious.

Reply:
Maybe you're right. I'll poke the code and see what I come up with.

Reply:
I opted not to store any CancellationErrors which then simplifies the code somewhat.